Use Binder build API, remove trigger_binder.py and selenium dependency, and add Binder CI stage

matthewfeickert opened the pull request:
# Description

Resolves #382 

Waiting on #380 and #381 to be merged first.

# Checklist Before Requesting Reviewer

- [x] Tests are passing
- [x] "WIP" removed from the title of the pull request
- [x] Selected an Assignee for the PR to be responsible for the log summary

# Before Merging

For the PR Assignees:

- [x] Summarize commit messages into a comprehensive review of the PR

```
* Use the Binder build API endpoint with Python's built in webbrowser to trigger builds
   - https://gitter.im/jupyterhub/binder?at=5c2f87038dafa715c73ff54f
   - https://github.com/jupyterhub/binderhub/blob/9ca8fa68bb8b69c6a2736f2275279583073f314f/examples/binder-api.py#L28
* Remove selenium and chromium-chromedriver dependency installation from Travis CI docs stage
* Remove binder/trigger_binder.py as being able to access the API makes it unnecessary
* Remove redundant Binder build API calls
* Move Binder build API call to separate Binder stage that runs only on PR merges to master
```

minrk commented on .travis.yml:
```diff
@@ -56,30 +56,22 @@ jobs:
-      - python binder/trigger_binder.py --url https://mybinder.org/v2/gh/diana-hep/pyhf/master?filepath=docs%2Fexamples%2Fnotebooks%2Fbinderexample%2FStatisticalAnalysis.ipynb
+      # Use Binder build API to trigger repo2docker to build image
+      - python -m webbrowser "https://mybinder.org/build/gh/diana-hep/pyhf/master"
+      - python -m webbrowser "https://mybinder.org/build/gh/diana-hep/pyhf/master?filepath=docs%2Fexamples%2Fnotebooks%2Fbinderexample%2FStatisticalAnalysis.ipynb"
```
FWIW, filepath isn't part of the build, it is only consumed by the browser redirects after build completes. These two lines won't do different things.

Additionally, you might want to use travis [environment variables](https://docs.travis-ci.com/user/environment-variables/#default-environment-variables) to ensure that the PR is being tested. I *think* that `https://mybinder.org/build/gh/$TRAVIS_REPO_SLUG/$TRAVIS_COMMIT` will be right. You *might* need to use $TRAVIS_PULL_REQUEST_SLUG for PRs since I believe that PR commits are available as refs in the target repo even before it's merged. Note that binder builds are cached by commit sha (it resolves branches and tags to a commit before checking), so building `abc123` and then requesting `master` later will use the cached build as long as the commits match.

Alternately, if this isn't really a test and you only want to build master, moving it to the `deploy` stage might be appropriate, since that only runs on master, not PRs.

matthewfeickert replied:
@minrk thanks for the feedback!

>  filepath isn't part of the build, it is only consumed by the browser redirects after build completes. These two lines won't do different things.

Ah, so if I understand correctly, you're saying that I have been misunderstanding how Binder works and if I have Binder build an image off of the `master` branch then that image will be used for all subsequent requests off of that branch _regardless_ of filepaths. This makes sense now that I think about it, but I think I must have had some confusion in the past that if I requested a URL with `filepath=` in it then this actually built a *different* image then if `filepath=` wasn't there.

> Alternately, if this isn't really a test and you only want to build master, moving it to the `deploy` stage might be appropriate, since that only runs on master, not PRs.

This is a good point, as this is really what we're doing.

minrk replied:
> if I have Binder build an image off of the master branch then that image will be used for all subsequent requests off of that branch regardless of filepaths

Correct. Binder has two stages: build and launch. The inputs to the build are only the repo and ref, while the 'filepath' is an input to launch, indicating what URL to redirect users to when the server is ready.

matthewfeickert commented:
@lukasheinrich Let's actually hold off on this one for now

matthewfeickert commented:
@lukasheinrich @kratsg Given what @minrk has pointed out about how I've been adding redundant Binder build calls and the point about having it get called every time in the docs is unnecessary if we just want it to be built for every PR I've added a `binder` stage that like `benchmark` is only run on PR merges to `master`. I added a new stage as `deploy` only gets run on tags. Sound okay to you?

kratsg commented:
> @lukasheinrich @kratsg Given what @minrk has pointed out about how I've been adding redundant Binder build calls and the point about having it get called every time in the docs is unnecessary if we just want it to be built for every PR I've added a `binder` stage that like `benchmark` is only run on PR merges to `master`. I added a new stage as `deploy` only gets run on tags. Sound okay to you?

This is fine with me if (unless I've misunderstood @minrk ) we're guaranteed that the binder builds won't ever fail when the rest of our tests pass correctly.

kratsg commented on .travis.yml:
```diff
@@ -88,6 +79,14 @@ jobs:
+  - stage: binder
+    python: '3.6'
+    before_install: skip
+    install: skip
```
are we sure the script will get called if `install: skip` is used? We had this issue before where it might have skipped the rest of the stages...

matthewfeickert replied:
Good question. Luckily the answer is yes. :+1:  I tested this in a test branch and you can [see the Travis build here](https://travis-ci.org/diana-hep/pyhf/builds/476949260). You can see from the [`binder` stage](https://travis-ci.org/diana-hep/pyhf/jobs/476949265) that the call get executed properly and from [the config](https://travis-ci.org/diana-hep/pyhf/jobs/476949265/config) you can see that it used the following

```
  - stage: binder
    python: '3.6'
    before_install: skip
    install: skip
    script:
      # Use Binder build API to trigger repo2docker to build image
      - python -m webbrowser "https://mybinder.org/build/gh/diana-hep/pyhf/master"
    after_success: skip
```

matthewfeickert commented:
> we're guaranteed that the binder builds won't ever fail when the rest of our tests pass correctly.

Correct. As the `binder` stage is the last to run then for the build to fail `python -m webbrowser` would need to die or BinderHub would need to be down.
